src/query.c

```diff
@@ -441,20 +444,71 @@ static IndexIterator *iterateExpandedTerms(QueryEvalCtx *q, Trie *terms, const c
   return NewUnionIterator(its, itsSz, q->docTable, 1, opts->weight, type, str);
 }
 
+typedef struct {
+  IndexIterator **its;
+  size_t nits;
+  size_t cap;
+  QueryEvalCtx *q;
+  QueryNodeOptions *opts;
+  double weight;
+} ContainsCtx;
+
+static int rangeIterCb(const rune *r, size_t n, void *p);
+static int suffixIterCb(const char *s, size_t n, void *p);
+
 /* Ealuate a prefix node by expanding all its possible matches and creating one big UNION on all
- * of them */
+ * of them.
+ * Used for Prefix, Contains and suffix nodes.
+*/
 static IndexIterator *Query_EvalPrefixNode(QueryEvalCtx *q, QueryNode *qn) {
   RS_LOG_ASSERT(qn->type == QN_PREFIX, "query node type should be prefix");
 
   // we allow a minimum of 2 letters in the prefx by default (configurable)
   if (qn->pfx.tok.len < RSGlobalConfig.minTermPrefix) {
     return NULL;
   }
-  Trie *terms = q->sctx->spec->terms;
+  
+  IndexSpec *spec = q->sctx->spec;
+  Trie *t = spec->terms;
+  ContainsCtx ctx = {.q = q, .opts = &qn->opts};
 
-  if (!terms) return NULL;
+  if (!t) {
+    return NULL;
+  }
+
+  rune *str = NULL;
+  size_t nstr;
+  if (qn->pfx.tok.str) {
+    str = strToFoldedRunes(qn->pfx.tok.str, &nstr);
+  }
+
+  ctx.cap = 8;
+  ctx.its = rm_malloc(sizeof(*ctx.its) * ctx.cap);
+  ctx.nits = 0;
+
+  // spec support contains queries
+  if (spec->suffix) {
+    // all modifier fields are supported
+    if (qn->opts.fieldMask == RS_FIELDMASK_ALL ||
+       (spec->suffixMask & qn->opts.fieldMask) == qn->opts.fieldMask) {
+    Suffix_IterateContains(spec->suffix->root, str, nstr, qn->pfx.prefix,
+                           suffixIterCb, &ctx);
+    } else {
+      QueryError_SetErrorFmt(q->status, QUERY_EGENERIC, "Contains query on fields without WITHSUFFIXTRIE support");
+    }
+  } else {
+    TrieNode_IterateContains(t->root, str, nstr, qn->pfx.prefix, qn->pfx.suffix,
```

Comment:
Where is the timeout is implemented? How do you handle OnTimeout=Error (will it return error)?

Reply:
Ohh I see you use `maxPrefixExpansions` so I guess not timeout here, maybe the name `maxPrefixExpansions` should be renamed?

Reply:
Ok, taking it back because I do see you cover timeout... can you add some explanation on how the timeout works and how you respect OnTImeout policy?

Reply:
How about `maxAffixExpansions` or `maxAutoUnionExpansions`?

Reply:
`maxAffixExpansions` sounds OK, if we change the actual config name we should make sure to also keep the old configuration for backward compatibility.